src/storage/src/client/upload_object_buffered.rs

```diff
@@ -730,6 +790,107 @@ async fn handle_start_resumable_upload_response(response: reqwest::Response) ->
     location.to_str().map_err(Error::deser).map(str::to_string)
 }
 
+async fn next_chunk<T>(
+    payload: &mut InsertPayload<T>,
+    remainder: Option<bytes::Bytes>,
+    target_size: usize,
+) -> Result<(VecDeque<bytes::Bytes>, Option<bytes::Bytes>)>
+where
+    T: StreamingSource,
+{
+    let mut partial = VecDeque::new();
+    let mut size = 0;
+    if let Some(mut b) = remainder {
+        match b.len() {
+            n if n > target_size => {
+                let remainder = b.split_off(target_size);
+                partial.push_back(b);
+                return Ok((partial, Some(remainder)));
+            }
+            n if n == target_size => {
+                partial.push_back(b);
+                return Ok((partial, None));
+            }
+            _ => {
+                size += b.len();
+                partial.push_back(b);
+            }
+        }
+    }
+
+    while let Some(mut b) = payload.next().await.transpose().map_err(Error::io)? {
```

Comment:
optional: there is probably an opportunity to unify the loop with the above... but the `async` makes it challenging.

I wanted to say something like `while let Some(mut b) = remainder.or_else(|| payload.next()...)` and reset remainder if we use it. but that won't work.

Reply:
Nice. I spent way too long trying to figure it out myself, then gave up.